codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamChannelBootstrap.java

```diff
@@ -0,0 +1,237 @@
+import io.netty.channel.ChannelOption;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+import io.netty.util.AttributeKey;
+import io.netty.util.internal.UnstableApi;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static java.util.Collections.unmodifiableMap;
+
+/**
+ * A class that makes it easy to bootstrap a new HTTP/2 stream as a {@link Channel}.
+ *
+ * <p>The bootstrap requires a registered parent {@link Channel} with a {@link ChannelPipeline} that contains the
+ * {@link Http2MultiplexCodec}.
+ *
+ * <p>A child channel becomes active as soon as it is registered to an eventloop. Therefore, an active channel does not
+ * map to an active HTTP/2 stream immediately. Only once a {@link Http2HeadersFrame} has been sent or received, does
+ * the channel map to an active HTTP/2 stream. In case it was not possible to open a new HTTP/2 stream (i.e. due to
+ * the maximum number of active streams being exceeded), the child channel receives an exception indicating the reason
+ * and is closed immediately thereafter.
+ *
+ * <p>This class is thread-safe.
+ */
+// TODO(buchgr): Should we deliver a user event when the stream becomes active? For all stream states?
+@UnstableApi
+public class Http2StreamChannelBootstrap {
+
+    private volatile Channel parentChannel;
+    private volatile Http2MultiplexCodec multiplexCodec;
+    private volatile ChannelHandler handler;
+    private volatile EventLoopGroup group;
+    private final Map<ChannelOption<?>, Object> options;
+    private final Map<AttributeKey<?>, Object> attributes;
+
+    // Lock for parentChannel and multiplexCodec
+    private final Object lock = new Object();
+
+    public Http2StreamChannelBootstrap() {
+        options = new LinkedHashMap<ChannelOption<?>, Object>();
+        attributes = new LinkedHashMap<AttributeKey<?>, Object>();
+    }
+
+    // Copy constructor
+    Http2StreamChannelBootstrap(Http2StreamChannelBootstrap bootstrap0) {
+        synchronized (bootstrap0.lock) {
+            parentChannel = bootstrap0.parentChannel;
+            multiplexCodec = bootstrap0.multiplexCodec;
+        }
+        handler = bootstrap0.handler;
+        group = bootstrap0.group;
+        synchronized (bootstrap0.options) {
+            options = new LinkedHashMap<ChannelOption<?>, Object>(bootstrap0.options);
+        }
+        synchronized (bootstrap0.attributes) {
+            attributes = new LinkedHashMap<AttributeKey<?>, Object>(bootstrap0.attributes);
+        }
+    }
+
+    /**
+     * Creates a new channel that will eventually map to a local/outbound HTTP/2 stream.
+     */
+    public ChannelFuture connect() {
+        return connect(-1);
+    }
+
+    /**
+     * Used by the {@link Http2MultiplexCodec} to instantiate incoming/remotely-created streams.
+     */
+    ChannelFuture connect(int streamId) {
```

Comment:
I wonder if we should not make this consistent with `Bootstrap` and so accept `SocketAddress`. We could then have a special `StreamSocketAddress` to pass in.